packages/hebao_v1/contracts/modules/core/ForwarderModule.sol

```diff
@@ -72,10 +73,19 @@ contract ForwarderModule is BaseModule
             "INVALID_DESTINATION"
         );
 
-        // If a non-zero txAwareHash is provided, we do not verify signature against
-        // the `data` field. The actual function call in the real transaction will have to
-        // check that txAwareHash is indeed valid.
-        bytes memory data_ = (txAwareHash == 0) ? data : bytes("");
+        bytes memory data_ = data;
+
+        if (txAwareHash != 0) {
+            // If a non-zero txAwareHash is provided, we do not verify signature against
+            // the `data` field. The actual function call in the real transaction will have to
+            // check that txAwareHash is indeed valid.
+            require(nonce == 0, "NONCE_MUST_BE_0");
+            controller.hashStore().verifyAndUpdate(txAwareHash);
```

Comment:
We have to watch out here that the txAwareHash hash value doesn't get marked as used if the meta tx fails. Otherwise the relayer can still provide invalid data (but without getting a reimbursement), and the txAwareHash hash value that could be used with valid data would become unusable before it's actually used correctly.

So I think we either:
- allow the meta tx call to fail (current behaviour), then we need to do this in the meta tx call or only call `verifyAndUpdate` if the meta tx call is successful
- we don't allow the meta tx call to fail so that the complete Ethereum tx is reverted

Reply:
Doing `verifyAndUpdate` in the meta tx call doesn't work because it currently also acts as replay protection on the meta tx level if it's used, as you're probably aware as you moved it. :)

Reply:
Let me think about it:)